django/db/models/fields/related_descriptors.py

```diff
@@ -1194,6 +1200,47 @@ def get_prefetch_querysets(self, instances, querysets=None):
                 False,
             )
 
+        @property
+        def constrained_target(self):
+            # If the through relation's target field's foreign integrity is
+            # enforced, the query can be performed solely against the through
+            # table as the INNER JOIN'ing against target table is unnecessary.
+            if not self.target_field.db_constraint:
+                return None
+            db = router.db_for_read(self.through, instance=self.instance)
+            if not connections[db].features.supports_foreign_keys:
+                return None
+            hints = {"instance": self.instance}
+            manager = self.through._base_manager.db_manager(db, hints=hints)
+            filters = {self.source_field_name: self.instance.pk}
+            # Nullable target rows must be excluded as well as they would have
+            # been filtered out from an INNER JOIN.
+            if self.target_field.null:
+                filters["%s__isnull" % self.target_field_name] = False
+            return manager.filter(**filters)
+
+        def exists(self):
+            constrained_target = self.constrained_target
+            if (
+                constrained_target is not None
+                and superclass is Manager
+                and self.get_prefetch_cache() is None
+            ):
+                return constrained_target.exists()
```

Comment:
The only thing I would change here is the order or predicates from cheaper to most costly.

`superclass is Manager` is super cheap as it basically compares the memory address of both objects and `self.get_prefetch_cache() is None` is cheaper than `self.constrained_target` so I would order them as follow to avoid unnecessary calls to `constrained_target`.

```suggestion
            if (
                superclass is Manager
                and self.get_prefetch_cache() is None
                and (constrained_target := self.constrained_target) is not None                
            ):
                return constrained_target.exists()
```

The same can be done for `count()`.

Reply:
Thanks for the explanation on chaining of `.all()`. There's a lot more going on than I realized.

Thanks for pointing out this optimization! I've made the changes to include your suggestion.